rpc/client/httpclient.go

```diff
@@ -204,108 +204,101 @@ type WSEvents struct {
 	endpoint string
 	ws       *rpcclient.WSClient
 
-	subscriptions map[string]chan<- interface{}
 	mtx           sync.RWMutex
-
-	// used for signaling the goroutine that feeds ws -> EventSwitch
-	quit chan bool
-	done chan bool
+	subscriptions map[string]chan<- interface{}
 }
 
 func newWSEvents(remote, endpoint string) *WSEvents {
 	wsEvents := &WSEvents{
 		endpoint:      endpoint,
 		remote:        remote,
-		quit:          make(chan bool, 1),
-		done:          make(chan bool, 1),
 		subscriptions: make(map[string]chan<- interface{}),
 	}
 
 	wsEvents.BaseService = *cmn.NewBaseService(nil, "WSEvents", wsEvents)
 	return wsEvents
 }
 
-// Start is the only way I could think the extend OnStart from
-// events.eventSwitch.  If only it wasn't private...
-// BaseService.Start -> eventSwitch.OnStart -> WSEvents.Start
-func (w *WSEvents) Start() error {
-	ws := rpcclient.NewWSClient(w.remote, w.endpoint, rpcclient.OnReconnect(func() {
+func (w *WSEvents) OnStart() error {
+	w.ws = rpcclient.NewWSClient(w.remote, w.endpoint, rpcclient.OnReconnect(func() {
 		w.redoSubscriptions()
 	}))
-	err := ws.Start()
-	if err == nil {
-		w.ws = ws
-		go w.eventListener()
+	err := w.ws.Start()
+	if err != nil {
+		return err
 	}
-	return err
-}
 
-// Stop wraps the BaseService/eventSwitch actions as Start does
-func (w *WSEvents) Stop() error {
-	// send a message to quit to stop the eventListener
-	w.quit <- true
-	<-w.done
-	w.ws.Stop()
-	w.ws = nil
+	go w.eventListener()
 	return nil
 }
 
-func (w *WSEvents) Subscribe(ctx context.Context, query string, out chan<- interface{}) error {
-	if ch := w.getSubscription(query); ch != nil {
-		return errors.New("already subscribed")
+// Stop wraps the BaseService/eventSwitch actions as Start does
+func (w *WSEvents) OnStop() {
+	err := w.ws.Stop()
+	if err != nil {
+		w.Logger.Error("failed to stop WSClient", "err", err)
 	}
+}
 
-	err := w.ws.Subscribe(ctx, query)
+func (w *WSEvents) Subscribe(ctx context.Context, subscriber string, query tmpubsub.Query, out chan<- interface{}) error {
+	q := query.String()
+
+	err := w.ws.Subscribe(ctx, q)
 	if err != nil {
-		return errors.Wrap(err, "failed to subscribe")
+		return err
 	}
 
 	w.mtx.Lock()
-	w.subscriptions[query] = out
+	// subscriber param is ignored because Tendermint will override it with
+	// remote IP anyway.
+	w.subscriptions[q] = out
 	w.mtx.Unlock()
 
 	return nil
 }
 
-func (w *WSEvents) Unsubscribe(ctx context.Context, query string) error {
-	err := w.ws.Unsubscribe(ctx, query)
+func (w *WSEvents) Unsubscribe(ctx context.Context, subscriber string, query tmpubsub.Query) error {
+	q := query.String()
+
+	err := w.ws.Unsubscribe(ctx, q)
 	if err != nil {
 		return err
 	}
 
 	w.mtx.Lock()
-	defer w.mtx.Unlock()
-	ch, ok := w.subscriptions[query]
+	ch, ok := w.subscriptions[q]
 	if ok {
 		close(ch)
-		delete(w.subscriptions, query)
+		delete(w.subscriptions, q)
 	}
+	w.mtx.Unlock()
```

Comment:
Why not keep the defer? I know there's a small performance price but probably worth it for maintainability